Update request function on the REST client

radu-matei opened the pull request:
This commit updates the `req` to include the option to pass query params
and headers.
It also adds default methods to some structs to simplify initialization.

Commit:
Update request function on the REST client

This commit updates the `req` to include the option to pass query params
and headers.
It also adds default methods to some structs to simplify initialization.

Signed-off-by: Radu M <root@radu.sh>

radu-matei commented on src/rest.rs:
```diff
@@ -36,29 +37,71 @@ impl Client {
+        url: String,
+        method: Method,
+        body: Option<&T>,
+        query: Option<&Q>,
```
`Q: Serialize + ?Sized` is the type constraint that `reqwest` introduces for query parameters.

thomastaylor312 commented on src/projects.rs:
```diff
@@ -22,6 +22,21 @@ pub struct Project {
 }
 
+impl Project {
+    pub fn default(id: String, description: String, script: String) -> Self {
```
This should be called `new`. Generally if you have an actual default method, it will be implementing the `Default` trait. If any config is needed, `new` is recommended

thomastaylor312 commented on src/rest.rs:
```diff
@@ -1,6 +1,6 @@
 use crate::error::BrigadeError;
-use hyper::Method;
-use reqwest::Response;
+use hyper::{HeaderMap, Method};
```
These types are reexported in `reqwest`

thomastaylor312 commented on src/rest.rs:
```diff
@@ -36,29 +37,71 @@ impl Client {
         url: String,
         method: Method,
         body: Option<&T>,
+        query: Option<&Q>,
```
This feels odd. Consider switching all of these options to a `RequestBuilder`-like type. That way you can do:

```rust
client.request(url, method).body(body).query(query).headers(headermap).execute()
```

This would allow you to drop the options in the external API and handle them purely internally

thomastaylor312 commented on src/rest.rs:
```diff
@@ -36,29 +37,71 @@ impl Client {
         Ok(res)
     }
+
+    pub async fn req_with_basic_auth<T: Serialize + ?Sized, Q: Serialize + ?Sized>(
```
Generally, you have a `Client` builder (different from the request builder) that allows you to configure auth, https, etc. Consider having `basic_auth` and `root_auth` methods on that builder that returns a `Client`, particularly with how big this is getting. It also allows a user to configure any default headers they need to send (if you expose that method)

thomastaylor312 commented on src/worker.rs:
```diff
@@ -18,6 +18,24 @@ pub struct WorkerSpec {
 }
 
+impl WorkerSpec {
+    pub fn default(script: String) -> Self {
```
Same here with default

Commit:
Rename default functions to new

Signed-off-by: Radu M <root@radu.sh>

Commit:
Use `reqwest::RequestBuilder` in REST client

This commit removes all custom request building logic from the REST
client and simply reuses the `reqwest::RequestBuilder` struct.

Signed-off-by: Radu M <root@radu.sh>

Commit:
Remove unused comments

Signed-off-by: Radu M <root@radu.sh>

radu-matei commented:
Ended up removing all custom request building logic in favor of returning a new `reqwest::RequestBuilder`, which will can be automatically populated with defaults when constructing it in the client (like user agent, or existing token).

Thanks a lot for the feedback, @thomastaylor312!

thomastaylor312 commented on src/projects.rs:
```diff
@@ -22,6 +22,21 @@ pub struct Project {
+                event_subscriptions: None,
+                worker_template: WorkerSpec::new(script),
+            },
+            kubernetes: None,
```
Nit: If I remember some of the other code, this gets filled out later. Would it be difficult to do it here and remove the `Option`?

thomastaylor312 commented on src/projects.rs:
```diff
@@ -94,15 +116,11 @@ impl ProjectsClient {
         authn::Token,
-        meta::ObjectMeta,
-        projects::{Project, ProjectSpec, ProjectsClient},
+        projects::{Project, ProjectsClient},
```
Nit: Generally you see `super::*` in test mods as it simplifies the imports you need

thomastaylor312 commented on src/rest.rs:
```diff
@@ -1,22 +1,20 @@
 }
 
 impl ClientConfig {
-    pub fn default() -> Self {
```
Nit: so in this case, default is fine as you are passing no parameters (clippy might even tell you this if you run it, but I can't remember). If you are planning on allowing people to pass params to this in the future, `new` would be correct. However, I would recommend rolling up this and the params in the `Client::new` function to be a builder. So if you are specifying all the options:

```rust
let client = ClientBuilder::new("http://foo.bar").allow_insecure_connections().token("deadbeef").build();
```

Basic auth or oauth would look like this (and could do the auth flow for you):

```rust
let client = ClientBuilder::new("https://foo.bar").auth(user, pass).build();
```

And root auth would look like this:

```rust
let client = ClientBuilder::new("https://foo.bar").root_auth(pass).build();
```

Those last auth options would move the auth stuff into the top level client where it can be reused for all clients. However, all of this is probably best done in a follow up

thomastaylor312 commented on src/rest.rs:
```diff
@@ -1,22 +1,20 @@
+#[derive(Debug, Clone)]
 pub struct Client {
     pub config: ClientConfig,
     pub address: String,
```
I can't comment on the line below, but I would unexport the `rest` data member as you reexpose only the functionality you need within the `req` function

thomastaylor312 commented on src/rest.rs:
```diff
@@ -36,29 +34,16 @@ impl Client {
-        Ok(res)
+    pub fn req<U: IntoUrl>(&self, method: Method, url: U) -> RequestBuilder {
+        let req = self.rest.request(method, url);
+        match self.token.clone() {
```
Nit: to avoid an unneeded clone, you could do `self.token.as_ref()` and then do `t.clone()` only when you need it

thomastaylor312 reviewed (approved):
Pretty much all nits left, so LGTM

Commit:
Make REST client private in machinery

Signed-off-by: Radu M <root@radu.sh>
